### ISSUE TITLE
Do not filter when asked to sort

### ISSUE THREAD

**Issue by alex-mashin** [SemanticMediaWiki/SemanticMediaWiki#874]
# Overview

The `{{#ask:}}` query removes all results without value for a property that is specified in the "sort" parameter.
# Steps to Reproduce

For example, if I have a page P1 in category C with a property `[[has X::value]]` and another page P2 in same category C without the "has X" property. They both use the same template and same semantic form.
# Actual Results

When I write:

```
{{#ask: [[Category:C]]
  | ?has X
  | sort=has X
  | order=asc
}}
```

it displays only page P1.

But:

```
{{#ask: [[Category:C]]
  | ?has X
}}
```

displays both P1 and P2 pages.
# Expected Results

it should have displayed both P1 and P2 in both cases.

Reproduced from https://phabricator.wikimedia.org/T24583

**Comment by mwjames** [comment]
See also #540

**Comment by s7eph4n** [comment]
I think #540 is different. In the #540 scenario pages are excluded from the result set if they have a property, but that property points to a non-existent page. Here in the #874 scenario pages are excluded, that do not have the property at all.

And for what it's worth, this bug her (#874) is an old issue that comes up again and again but does not have a (formal) solution: You just can not sort what you do not have. In a list of cities sorted by area, where do you put those, that do not have the "Has area" property set? So IMHO it is not a bug.

**Comment by mwjames** [comment]
> I think #540 is different. In the #540 scenario pages are excluded from the result set if they have a property, but that property points to a non-existent page.

This is OK but I'd like to keep them connected since both are touching the `sort` parameter.

> So IMHO it is not a bug.

:-)

**Comment by alex-mashin** [comment]
I think, it *is* a bug, for all practical purposes. Pages are not supposed to be missing from query results if even some printouts are empty, and sorting property is even less important than a printout. Usually sorting is meant to bring some data closer to user's attention but not to remove other data completely.

In a SQL query, NULL is a sortable value.

In believe, absent properties should behave like zeroes and empty strings when sorted.

**Comment by s7eph4n** [comment]
> In believe, absent properties should behave like zeroes and empty strings when sorted.

Please sort by birth date:
* Lucan, 39 - 65
* Ovid, 43 B.C. - 17
* Ptolemy, 90 - 168
* Theocritus
* Virgil, 70 B.C. - 19 B.C.
* Homer, ~850 B.C.

**Comment by alex-mashin** [comment]
Homer, Virgil, Ovid, Lucal, Ptolemy, Theocritus (absent date == not yet == a date infinitely far in the future).

Alternatively, set default values in printout and sort statements in `{{#ask:}}`: like `?Date of birth#ISO~01.01.9999=Born` and `sort=Born, Name|order=asc, asc|sort_nulls=01.01.9999,(anonymous)`.

**Comment by s7eph4n** [comment]
> Homer, Virgil, Ovid, Lucal, Ptolemy, Theocritus (absent date == not yet == a date infinitely far in the future).

That's different from what you wanted to do before, which would have resulted in Homer, Virgil, Ovid, Theocritus, Lucan, Ptolemy.

Which is exactly my point: While it may seem obvious, what to do with the non-existent value for every specific case, it is not at all obvious for the general case.

However, I do agree that the returned result is problematic. Given the structure of the `#ask` query, I think the formally more correct thing to do would be to throw an error saying that the result set can not be sorted.
Why? You specify the selection criteria and the sorting criteria separately (`{{#ask:[[SomeProp::Foo]] |sort=SomeOtherProp}}`), which implies that you want to first select pages and then apply sorting to them. If sorting the selected pages is not possible an error should be returned.

A better solution could be to introduce some further allowed values than just `asc` or `desc` for the `order` parameter. This would also fix the issue that SMW today is assuming some inherent "natural" order for properties, which is already problematic.

**Comment by alex-mashin** [comment]
I didn't mention dates before you asked; I meant only strings and numbers. Any assumption about missing dates is better than dropping rows altogether. Nil cast to date type, plus infinity, specially set default value for the query, anything.

And I don't think that making a correctly written query fail, whatever the data be, is a good idea.

The point is, in a production semantic wiki, incomplete data is a typical case, not an exception, and should not be treated like an error. The software ought to try to make most of it.

**Comment by mwjames** [comment]
> A better solution could be to introduce some further allowed values than just asc or desc for the order parameter. This would also fix the issue that SMW today is assuming some inherent "natural" order for properties, which is already problematic.

I turned your comment into a separate RFC #902 to avoid losing it in the dust of this discussion. I won't have time to spent on it but it would be marvelous if you could extent the RFC with whatever ideas you have in mind.

**Comment by Rudloff** [comment]
Any news on this? I still hit this issue pretty often when dealing with non-mandatory properties and our users are frequently confused as to why some pages don't appear in sorted lists.

**Comment by mwjames** [comment]
refs #2823

**Comment by ankostis** [comment]
For future reference, #2823, merged in 3.0.0, solves this if the following is added in th `LocalSettings.php`:

    $GLOBALS['smwgQSortFeatures'] = $GLOBALS['smwgQSortFeatures'] | SMW_QSORT_UNCONDITIONAL;
